packages/legend-query/src/components/QueryBuilderCustomDatePicker.tsx

```diff
@@ -0,0 +1,1267 @@
+          );
+      }
+    }
+    case QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.FIRST_DAY_OF_YEAR: {
+      const firstDayOfYearSFE = new SimpleFunctionExpression(
+        SUPPORTED_FUNCTIONS.FIRST_DAY_OF_YEAR,
+        multiplicityOne,
+      );
+      firstDayOfYearSFE.genericType = GenericTypeExplicitReference.create(
+        new GenericType(date),
+      );
+      return firstDayOfYearSFE;
+    }
+    case QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.FIRST_DAY_OF_QUARTER: {
+      const firstDayOfQuarterSFE = new SimpleFunctionExpression(
+        SUPPORTED_FUNCTIONS.FIRST_DAY_OF_QUARTER,
+        multiplicityOne,
+      );
+      firstDayOfQuarterSFE.genericType = GenericTypeExplicitReference.create(
+        new GenericType(strictDate),
+      );
+      return firstDayOfQuarterSFE;
+    }
+    case QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.FIRST_DAY_OF_MONTH: {
+      const firstDayOfMonthSFE = new SimpleFunctionExpression(
+        SUPPORTED_FUNCTIONS.FIRST_DAY_OF_MONTH,
+        multiplicityOne,
+      );
+      firstDayOfMonthSFE.genericType = GenericTypeExplicitReference.create(
+        new GenericType(date),
+      );
+      return firstDayOfMonthSFE;
+    }
+    case QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.FIRST_DAY_OF_WEEK: {
+      const firstDayOfWeekSFE = new SimpleFunctionExpression(
+        SUPPORTED_FUNCTIONS.FIRST_DAY_OF_WEEK,
+        multiplicityOne,
+      );
+      firstDayOfWeekSFE.genericType = GenericTypeExplicitReference.create(
+        new GenericType(date),
+      );
+      return firstDayOfWeekSFE;
+    }
+    case QUERY_BUILDER_CUSTOM_DATE_PICKER_OPTION.PREVIOUS_DAY_OF_WEEK: {
+      const previousFridaySFE = new SimpleFunctionExpression(
+        SUPPORTED_FUNCTIONS.PREVIOUS_DAY_OF_WEEK,
+        multiplicityOne,
+      );
+      previousFridaySFE.genericType = GenericTypeExplicitReference.create(
+        new GenericType(date),
+      );
+      const dayOfWeekEnumIntanceValue = new EnumValueInstanceValue(
+        GenericTypeExplicitReference.create(
+          new GenericType(graph.getType(DAY_OF_WEEK)),
+        ),
+        multiplicityOne,
+      );
+      dayOfWeekEnumIntanceValue.values.push(
+        EnumValueExplicitReference.create(
+          guaranteeNonNullable(
+            graph
+              .getEnumeration(DAY_OF_WEEK)
+              .values.filter(
+                (e) =>
+                  e.name ===
+                  (datePickerOption as CustomPreviousDayOfWeekOption).day,
+              )[0],
+          ),
+        ),
+      );
+      previousFridaySFE.parametersValues.push(dayOfWeekEnumIntanceValue);
+      return previousFridaySFE;
+    }
+    default:
+      throw new UnsupportedOperationError(
+        `${datePickerOption.value} is not supported yet.`,
+      );
+  }
+};
+
+/**
+ * Generate the enum value of type Pure Enum, DURATION_UNIT, based on the input string.
+ */
+const generatePureDurationEnumValueFromString = (
```

Comment:
To make it easier to analyze and read the code, let's stick to the naming convention:

`valueSpec -> UI state` -> process...
`UI state -> valueSpec` -> build...